### PR TITLE
[WIP] Naturalspeech 2 Implementation 

### PR DESCRIPTION
**Pull request by manmay-nakhashi**
Paper --> https://arxiv.org/pdf/2304.09116.pdf

### CODE REVIEW

**Comment by kungfooman** [comment]
Thank you for the PR! I heard some of the Naturalspeech 2 examples and they sound great. But are there any models/weights to test/download this with?

Or is it possible to train our own language with this already?

**Comment by manmaynakhashi** [comment]
@kungfooman this is still work in progress since i can only work on weekends on this it'll take some time to complete this code, right now i am still resolving some errors.

**Comment by manmay-nakhashi** [comment]
@kungfooman this is still work in progress since i can only work on weekends on this it'll take some time to complete this code, right now i am still resolving some errors.

**Comment by manmay-nakhashi** [comment]
Status Update: 
- all the shapes are fixed train_naturalspeech2 is runnable ,  only data loss is added
- [TODO] sengemted training implementation 
-  [TODO] implement other two losses
- [TODO] add mel_loss and check if that helps

**Comment by Iamgoofball** [comment]
Let me know when you want this tested, I'd be happy to give it a run

**Comment by manmay-nakhashi** [comment]
> Let me know when you want this tested, I'd be happy to give it a run on a multispeaker use case

there are still few bugs in pitch and duration pipeline i have resolved it , but i need test it once and complete the inference function, till then it's still not trainable, once it starts generating voice on toy dataset i'll post it here.

**Comment by manmay-nakhashi** [comment]
@erogol score loss is not yet implemented , can you look at the forward function if it looks ok ? training script is running but inference is not complete.

File: TTS/tts/models/naturalspeech2.py [hunk 1]
```diff
@@ -0,0 +1,1223 @@
...
+        remaining_mask = torch.ones_like(latents, dtype=torch.bool)
+
+        # Get random segment for the speech prompt
+        speech_prompts, segment_indices = rand_segments(
+            latents, latents_lengths, self.diff_segment_size, let_short_samples=True, pad_short=True
+        )
+
+        # iterate over the batch dimension
+        for i in range(latents.size(0)):
+            remaining_mask[i, :, segment_indices[i] : segment_indices[i] + self.diff_segment_size] = 0
```
Comment (erogol) [inline TTS/tts/models/naturalspeech2.py:718]: isn't it easier to just iterate over the `remaning_latents`? `remaining_mask` is not being used somewhere else look like.

File: TTS/tts/models/naturalspeech2.py [hunk 2]
```diff
@@ -0,0 +1,1223 @@
...
+
+        # apply masks to get remaining parts
+        remaining_latents = latents.masked_select(remaining_mask).view(latents.shape[0], latents.shape[1], -1)
+        remaining_latents_lengths = torch.tensor(remaining_latents.shape[1:2]).to(remaining_latents.device)
+
+        # Encode speech prompt
+        speech_prompts_enc = self.prompt_encoder(speech_prompts)
```
Comment (erogol) [inline TTS/tts/models/naturalspeech2.py:725]: I'd make a separate function to compute the prompt and return the transposed tensor to get rid of the transposes below.

File: TTS/tts/models/naturalspeech2.py [hunk 3]
```diff
@@ -0,0 +1,1223 @@
...
+        alignment_hard = torch.sum(alignment_mas, -1).int()
+        alignment_soft = alignment_soft.squeeze(1).transpose(1, 2)
+        return alignment_hard, alignment_soft, alignment_logprob, alignment_mas
+
+    @staticmethod
+    def _set_cond_input(aux_input: Dict):
+        # [TODO] use this
+        return None
+
+    def forward(  # pylint: disable=dangerous-default-value
```
Comment (erogol) [inline TTS/tts/models/naturalspeech2.py:690]: looks like the first Q-K-V Attention is missing that is with random m embeddings in Fig3 in the paper

Comment (manmay-nakhashi) [reply TTS/tts/models/naturalspeech2.py:690]: i have added this in diffusion.py

File: TTS/tts/models/naturalspeech2.py [hunk 4]
```diff
@@ -0,0 +1,1223 @@
...
+        alignment_mas = alignment_mas.transpose(1, 2)
+
+        durations_pred = self.duration_predictor(phoneme_enc.transpose(1, 2), speech_prompts_enc.transpose(1, 2))
+        durations_pred = durations_pred.squeeze(1)
+
+        pitch = average_over_durations(pitch, alignment_hard)
+        pitch_pred = self.pitch_predictor(phoneme_enc.transpose(1, 2), speech_prompts_enc.transpose(1, 2))
+
+        durations_pred_exp = durations_pred.unsqueeze(2)
+        expanded_dur = durations_pred_exp.expand_as(phoneme_enc)
```
Comment (erogol) [inline TTS/tts/models/naturalspeech2.py:741]: shouldn't we expanding the `phoneme_enc` not `duraitons_pred`

File: TTS/tts/models/naturalspeech2.py (hunk 3 above)
Comment (manmay-nakhashi) [reply TTS/tts/models/naturalspeech2.py:690]: class Denoise

File: TTS/tts/models/naturalspeech2.py (hunk 4 above)
Comment (manmay-nakhashi) [reply TTS/tts/models/naturalspeech2.py:741]: duration pred shape is b , t , 1 and phoneme_enc size is b,t , 512 so here matching the size to match the dumentions, how can i expand phoneme_enc based on duration is not clear to me

File: TTS/tts/models/naturalspeech2.py (hunk 3 above)
Comment (erogol) [reply TTS/tts/models/naturalspeech2.py:690]: then no worries 👍

File: TTS/tts/models/naturalspeech2.py (hunk 4 above)
Comment (erogol) [reply TTS/tts/models/naturalspeech2.py:741]: this is what we do in VITS 

```python
from TTS.tts.utils.helpers import generate_path

....

# convert duraitons to attn mask 
attn_mask = x_mask * y_mask.transpose(1, 2)  # [B, 1, T_enc] * [B, T_dec, 1]
attn = generate_path(durations.squeeze(1), attn_mask.squeeze(1).transpose(1, 2))

...

 # expand prior
m_p = torch.einsum("klmn, kjm -> kjn", [attn, m_p])
logs_p = torch.einsum("klmn, kjm -> kjn", [attn, logs_p])
```

**Comment by manmay-nakhashi** [comment]
Update : code clean up is left , mostly model implementation is complete i am training on vctk and see if i am able to get some output.
i'll post the tensorboard over here.

**Comment by erogol** [comment]
I think it is better to implement synthesize in the model going forward. It'd be more flexible.

**Comment by erogol** [comment]
This is buggy -> https://github.com/coqui-ai/TTS/blob/755405d5ca5956dc073144c395332d1b24286cca/TTS/tts/models/naturalspeech2.py#LL757C65-L757C69

I think, you should be using the aligner attention not the predicted one while training

File: TTS/tts/models/naturalspeech2.py [hunk 5]
```diff
@@ -0,0 +1,1276 @@
...
+
+            wav = batch["wav"][i]
+            wav_padded[i, :, : wav.size(1)] = torch.FloatTensor(wav)
+
+            # latents_emb = batch["latents"][i]
+            # latents_padded[i, :, : latents_emb.size(2)] = torch.FloatTensor(latents_emb)
+        # Extract discrete codes from EnCodec
+        emb, codes, _ = self.encodec(wav_padded, return_encoded=True)
+        emb = emb.squeeze(1)
+        emb = emb.transpose(1, 2)
```
Comment (erogol) [inline TTS/tts/models/naturalspeech2.py:391]: If you do this in `format_batch_on_device` you can speed up the data loading significantly.

**Review by erogol — CHANGES_REQUESTED** [review]
